### ISSUE TITLE
Dumping Vars in templates for development

### ISSUE THREAD

**Issue by Drudwyn** [CloudyKit/jet#111]
For rapid cross-team development, it would be useful to allow a developer to dump Vars in a template in a pretty print, such as Go-Spew. Is this achievable?

**Comment by ryanlath** [comment]
Not pretty, but I use this:
```
views.AddGlobalFunc("dump", func(a jet.Arguments) reflect.Value {
	a.RequireNumOfArguments("dump", 1, 1)
	return reflect.ValueOf(fmt.Sprintf("%#v", a.Get(0)))
})
```

Then in template:
`{{ dump(VAR) }}`

**Comment by ryanlath** [comment]
I should really read the docs.  There's already a built in func:
`{{ json(VAR) }}`
Then, I'm sure there's a golang JSON prettifier out there.

**Comment by Dexus** [comment]
Would be nice to have a debug function, that returns all variables that are availabel in the context.

Don't like to use some variables that not there to read the error with the varaibels.

**Comment by annismckenzie** [comment]
Could you take a crack at that? It sure would be nice.

**Comment by sauerbraten** [comment]
I'd like dump() to be smart and versatile, so I imagine an API that allows the following calls:
```
dump()
dump(2)
dump("foo")
dump("foo", "bar", "lorem", "ipsum")
```

In the first case, I'd expect dump to print multiple sections:
1. everything in Runtime.context
2. everything in Runtime.variables
3. everything in Runtime.set.globals
3. everything in Runtime.blocks

In the second case, when called with a single int argument, dump should interpret that as the maximum depth of parent scopes to walk up, and additionally print Runtime.parent.variables and Runtime.parent.parent.variables between the Runtime.variables section and the Runtime.set.globals section. (This means the first case is just like the second case, just with an implied depth of 0.)

The third and fourth case are handled the same way: for a call with one or more string arguments, dump would treat each string as an identifier name and print the name, type, and value (and maybe even scope?) of whatever was found when looking up each identifier. The lookup depth in parent scopes is unlimited, like in normal jet usage. For example,
```
foo [string]:
  "bar"
```
or
```
foo(type="text", label, id, value="", required=false) [block]:
  <div class="form-field">
      <label for="{{ id }}">{{ label }}</label>
      <input type="{{ type }}" value="{{ value }}" id="{{ id }}" {{ required ? "required" : "" }} />
  </div>
```

So an identifier could be describe a variable, something from the context, a global, or a block. (Maybe it could even be a template path and dump the template AST?)

The format of each "thing" in the first and second calls could be the same as in the third and fourth call, or some abbreviated form.

To elegantly handle the first and second calls, there should exist dumpContext, dumpVariables, dumpGlobals and dumpBlocks functions, which might also be useful to export directly to the user... 🤔

**Comment by jan-herout** [comment]
OK, I think I've got someting which I think covers the first two cases, albeit not cleanly. Refactoring is needed.

Before I dive more into code details, perhaps you could provide me a few pointers. In the end, Jet is still a new codebase for me, and I suspect my primary background is dramatically different from yours - I am a data warehouse developer first, Go is one of the tools which I sometimes use.  I "do not code Go for a living".

Incidentally, I plan to use Jet on a project where it would help me automate generation of code (scripts) for the warehouse, so I am motivated (a) to understand Jet better, (b) to provide some value back to the wonderful people, who created this library, if it is within my capabilities.

Perhaps you could look at the following rant, and share your thoughts?

### "Panic policy"

The API you suggested assumes that the `dump` function is used with:
- no parameters
- exactly one `int` parameter
- one or more `string` parameters

What should happen if you violate these assumptions? Should it panic? Should it render an error into the template?

### Dump by identifier

As for dumping the template ATS, I feel this is for now beyond my capabilities. Perhaps it can start small, and this functionality can be provided later. For now, I feel like I should be able to:

- dump a variable defined in the template
- dump a variable provided to the template in the `VarMap` during `Execute` call
- dump a block defined in the template

Following could be provided later on:

- identifier pointing to a struct field within the context
- identifier pointing to a template, printing AST

### Lookup by ID

Is there already some method implemented, which - based on an identifier - would return either a member of `VarMap` (a variable), or a block? As I understand it, there is a need to go through the "current runtime" and all its parents, and to locate "the thing" by name in `variable` and `blocks`.

Do I have to implement such a lookup from the scratch? If so:

- Where do you feel like implementing it - would it be a method (methods) on `Runtime`? 
- Would it be exported, or unexported? I suppose unexported.
- Should it stop after it succeeds in locating the first "thing" by the name? (I suspect yes... not sure though.)


There might be more questions in the future. Sorry, and **thanks for your patience.**

**Comment by sauerbraten** [comment]
Great to hear you didn't lose interest after my previous, lengthy comment on this! We're always happy about people contributing, so ask away!

> What should happen if you violate these assumptions? Should it panic? Should it render an error into the template?

Since I would put a big disclaimer "FOR DEVELOPMENT ONLY" in front of such a dump() function, it's completely fine if we just panic when the arguments don't match what dump() expects.

> Dump by identifier

Variables passed in the VarMap when calling Execute() and variables declared in the demplate actually end up in the same place: [Runtime.variables (via the embedded Runtime.scope struct)](https://github.com/CloudyKit/jet/blob/4beb91506060bd8900ec0f7d42e401d21300350f/eval.go#L76), so dumping stuff in there should cover the first two things you want to do.

Blocks are in [Runtime.blocks](https://github.com/CloudyKit/jet/blob/4beb91506060bd8900ec0f7d42e401d21300350f/eval.go#L98), conveniently indexed by name, so iterating or looking them up should not be difficult.

> Do I have to implement such a lookup from the scratch?

For variables and globals, no: there is a resolve() function on the Runtime type: https://github.com/CloudyKit/jet/blob/4beb91506060bd8900ec0f7d42e401d21300350f/eval.go#L175

Looking up a block should be easy using the map mentioned before. (Blocks aren't scoped, but rather just copied over from the template that you extend from, so there's no need to look up blocks anywhere else.)

> Should it stop after it succeeds in locating the first "thing" by the name?

Good question! All variables/globals accessible in the current scope are uniquely named, so resolve() will only give you one value. I believe however that there can be a variable and a block, both named "foo", at the same time, so maybe it would be good to always look inside Runtime.blocks, even if resolve() already found something, and then to dump both.

> There might be more questions in the future. Sorry, and thanks for your patience.

Sure, please ask when you get stuck somewhere. And no worries, thank YOU for working on this!

**Comment by sauerbraten** [comment]
Added in #189 and released in [v6.1.0](https://github.com/CloudyKit/jet/releases/tag/v6.1.0).
